### PR TITLE
swap ed25519-dalek for ed25519-zebra

### LINKED ISSUE #6618: Bring back signature batch verification
https://github.com/paritytech/substrate/pull/6616 removed batch verification (the usage of it) as it introduced new host functions that are doing the batch verification.

We need to bring back the runtime implementation of the batch verification.

This requires a new trait:
```
/// A signature that supports batch verification.
trait BatchVerify: Verify {
    /// Register a signature for batch verification.
    ///
    /// This requires that batch verification is enabled by doing XYZ.
    ///
    /// Returns `true` when the signature was successfully registered for batch verification or if batch verification is not enabled
    /// the signature could be verified successfully immediately.
    ///
    /// # Warning
    /// 
    /// This requires that the batch verification is finished by calling XYZ to check the result of all submitted signature verifications.
    fn batch_verify<L: Lazy<[u8]>>(&self, msg: L, signer: &<Self::Signer as IdentifyAccount>::AccountId) -> bool;
}
``` 

This trait needs to be implemented for all the required types and we need to switch `UncheckedExtrinsic` to use this trait.

### LINKED ISSUE #8055: ed25519-zebra
Are we using the [prehashed](https://docs.rs/ed25519-dalek/1.0.1/ed25519_dalek/index.html?search=prehash) variant of ed25519 anyplace?  If not, then we [should](https://hdevalence.ca/blog/2020-10-04-its-25519am) replace ed25519-dalek with [ed25519-zebra](https://github.com/ZcashFoundation/ed25519-zebra), which better supports batch verification. 

Ain't clear if ed25519-zebra ever wants features not required by zcash, so maybe we should do something else if we use the prehashed variant of ed25519 somewhere.  I once pointed @tomaka to its existence

### PR DESCRIPTION
**Pull request by jakehemmerle**
polkadot companion: https://github.com/paritytech/polkadot/pull/3504
This should close #8055. All tests are pass in `sp-core`.

This should eventually close #6618, but the `batching_works` test in `sp-io` does not pass. Since this passes on the current Substrate master, this needs to get fixed.

`batching_works` stdout and panic log: https://pastebin.com/bA78t00a

@burdges maybe you can take a look at this?

### CODE REVIEW

**Comment by burdges** [comment]
You still have the failing test case?  If I recall, it looked like a case of using the identity as a key, likely zeros do not represent the identity but zebbra pushes them into being the identity or something.  Where was it?

**Comment by jakehemmerle** [comment]
> You still have the failing test case? If I recall, it looked like a case of using the identity as a key, likely zeros do not represent the identity but zebbra pushes them into being the identity or something. Where was it?

yes it's still failing. Failing case is `batching_works` in `sp-io`.

I didn't realize your identity key matrix message was from reading the code; I thought it was just an example!

**Comment by jakehemmerle** [comment]
Tests fixed! Should close #6618 and #8055.

One question: sr25519 (as did ed25519-dalek) throws an error when verifying identity arguments, while ed25519-zebra is fine with verifying identities. I've changed the test to expect this behavior; is this fine or will these differences be an issue in practice?

**Comment by burdges** [comment]
It depends what you mean by identity.  Ristretto has the same identity as [Ed25519](https://doc.dalek.rs/src/curve25519_dalek/edwards.rs.html#300), but serializes it as `[0u8; 32]`.  I think schnorrkel would verify an identity signature, except `[0u8; 64]` triggers [`SignatureError::NotMarkedSchnorrkel`](https://github.com/w3f/schnorrkel/blob/master/src/sign.rs#L129). 

[Ed25519](https://doc.dalek.rs/src/curve25519_dalek/edwards.rs.html#276) serializes its identity not as `[0u8; 32]`.  Yet, `[0u8; 32]` lies on the curve in the subgroup of order 8, so likely 1 in 8 messages would validate under the zero signature and public key.

**Comment by burdges** [comment]
I do not care too much how we handle public keys consisting of zero bytes.

**Comment by jakehemmerle** [comment]
I swapped verifications using zero keys with signatures from another keypair. This should fix the above issues mentioned by Jeff.

File: primitives/io/src/lib.rs [hunk 1]
```diff
@@ -1561,15 +1561,6 @@ mod tests {
...
 		ext.register_extension(TaskExecutorExt::new(TaskExecutor::new()));
 		ext.execute_with(|| {
-			// invalid ed25519 signature
-			crypto::start_batch_verify();
-			crypto::ed25519_batch_verify(
-				&Default::default(),
-				&Vec::new(),
-				&Default::default(),
-			);
-			assert!(!crypto::finish_batch_verify());
```
Comment (burdges) [inline primitives/io/src/lib.rs:1571]: I'd expect 1/8th of the possible messages makes this succeed, so this test was mostly useless anyways.

**Comment by burdges** [comment]
I think https://github.com/paritytech/substrate/blob/d717bdb9b432cb2295fddc8505665698e9cd3397/primitives/io/src/lib.rs#L1542 likely fails in deserialization, which works fine, but maybe duplicate it with a test that fails like https://github.com/paritytech/substrate/blob/d717bdb9b432cb2295fddc8505665698e9cd3397/primitives/io/src/lib.rs#L1588 does.

**Comment by burdges** [comment]
LGTM modulo adding one test.  :) 

We'll get someone else from parity to review though.

**Comment by jakehemmerle** [comment]
Sweet; final test fixed!

**Comment by bkchr** [comment]
@jakehemmerle can you please merge master

**Comment by jakehemmerle** [comment]
@bkchr merged

**Comment by bkchr** [comment]
Sorry for the late reply, but the CI wasn't happy about your merge.

**Comment by jakehemmerle** [comment]
Looks like the lockfile. Will update shortly

**Comment by jakehemmerle** [comment]
@bkchr I fixed the lockfile and it's now passing the test-linux-stable-int build, but still failing the test-linux-stable due to a pallet_ui.rs test. I just spend almost two hours trying to understand why it's failing, but I don't have enough context to know where to look and on the surface it looks unrelated (although I see other test-linux-stable builds passing, which means it's probably something I've done). I see a new trait_constant_valid_bounds.rs test has been added to pallet_ui by Andrew two weeks ago, maybe related? that new test is: https://github.com/paritytech/substrate/blob/master/frame/support/test/tests/pallet_ui.rs#L26

test-linux-stable failing at https://gitlab.parity.io/parity/substrate/-/jobs/987864#L988 

Do you think you can take a look and point me in the right direction?

**Comment by bkchr** [comment]
@jakehemmerle you just need to merge master.

**Comment by bkchr** [comment]
This should fix your problem

**Comment by jakehemmerle** [comment]
@bkchr I had been rebasing before; just merged, still having the same issue

**Comment by bkchr** [comment]
Sorry for all the delays!

@jakehemmerle could you please create a polkadot pr that uses your Substrate branch. I would like to run a burn in (syncing Kusama and Polkadot) to ensure that everything works.

**Comment by gilescope** [comment]
Sorry to make things harder @jakehemmerle in the short term, but we've switched on using rustfmt so you might want to merge master.

**Comment by bkchr** [comment]
> Sorry to make things harder @jakehemmerle in the short term, but we've switched on using rustfmt so you might want to merge master.

He already merged these changes as you see with the CI status.

File: primitives/core/src/ed25519.rs [hunk 2]
```diff
@@ -63,16 +63,18 @@ pub struct Public(pub [u8; 32]);
...
 	fn clone(&self) -> Self {
-		Pair(ed25519_dalek::Keypair {
-			public: self.0.public,
-			secret: ed25519_dalek::SecretKey::from_bytes(self.0.secret.as_bytes())
-				.expect("key is always the correct size; qed"),
-		})
+		Pair {
+			public: self.public.clone(),
+			secret: self.secret.clone(),
+		}
```
Comment (gilescope) [inline primitives/core/src/ed25519.rs:77]: We can just do #[derive(Clone)] on Pair now and remove this impl?

File: primitives/core/src/ed25519.rs [hunk 3]
```diff
@@ -484,10 +486,11 @@ impl TraitPair for Pair {
 	///
 	/// You should never need to use this; generate(), generate_with_phrase
 	fn from_seed_slice(seed_slice: &[u8]) -> Result<Pair, SecretStringError> {
-		let secret = ed25519_dalek::SecretKey::from_bytes(seed_slice)
+		// does try_into consume the seed? can I consume seed_slice here? I think not right?
```
Comment (gilescope) [inline primitives/core/src/ed25519.rs:489]: You can't consume the seed in this fn but have raised a PR to make from_seed (above) take ownership of the Seed.
https://github.com/paritytech/substrate/pull/9683

That then allows `SigningKey::from(seed)` in the from_seed function.

File: primitives/core/src/ed25519.rs [hunk 4]
```diff
@@ -508,15 +511,14 @@ impl TraitPair for Pair {
 
 	/// Get the public key.
 	fn public(&self) -> Public {
-		let mut r = [0u8; 32];
-		let pk = self.0.public.as_bytes();
-		r.copy_from_slice(pk);
-		Public(r)
+		// does this consume public? Is that why we used copy_from_slice?
```
Comment (gilescope) [inline primitives/core/src/ed25519.rs:514]: nope it won't consume public. - it can't take ownership of anything from an &self as it's read only.
(If it was a problem it wouldn't compile.)

File: primitives/core/src/ed25519.rs [hunk 5]
```diff
@@ -530,17 +532,20 @@ impl TraitPair for Pair {
...
+		let sig = match ed25519_zebra::Signature::try_from(sig) {
 			Ok(s) => s,
 			Err(_) => return false,
 		};
 
-		public_key.verify(message.as_ref(), &sig).is_ok()
+		match public_key.verify(&sig, message.as_ref()) {
+			Ok(_) => true,
+			_ => false,
+		}
```
Comment (gilescope) [inline primitives/core/src/ed25519.rs:548]:
```suggestion
		public_key.verify(&sig, message.as_ref()).is_ok()
```

File: primitives/io/src/lib.rs [hunk 6]
```diff
@@ -1668,12 +1666,13 @@ mod tests {
...
+			let pair1 = ed25519::Pair::generate_with_phrase(None).0;
+			let pair2 = ed25519::Pair::generate_with_phrase(None).0;
 			let msg = b"Important message";
-			let signature = pair.sign(msg);
-			crypto::ed25519_batch_verify(&signature, msg, &pair.public());
+			let signature = pair1.sign(msg);
 
-			crypto::ed25519_batch_verify(&Default::default(), &Vec::new(), &Default::default());
+			crypto::ed25519_batch_verify(&signature, msg, &pair1.public());
+			crypto::ed25519_batch_verify(&signature, msg, &pair2.public());
```
Comment (gilescope) [inline primitives/io/src/lib.rs:1675]: which one is invalid?

Comment (jakehemmerle) [reply primitives/io/src/lib.rs:1675]: The second line (1675) should now be invalid, but I realize I accidentally used a different key pair (`pair` instead of `pair1`) to generate the signature. Fixing now.

File: primitives/core/src/ed25519.rs (hunk 3 above)
Comment (jakehemmerle) [reply primitives/core/src/ed25519.rs:489]: I agree with your point in that PR that the seed should be consumed; will update this if that gets merged

**Comment by burdges** [comment]
We've seemingly never merged this.  Anything problematic crop up?  Can we get this running on one of the testnets?

**Comment by gilescope** [comment]
@bkchr is there someone on your team that might put this to bed? Might be up @wigy-opensource-developer's street? Zebra is harder to make mistakes with than Dalek - am keen if we can have fewer cryptographic footguns.

**Comment by bkchr** [comment]
Maybe at some point. Going to close this one as it is stale anyway.

**Comment by jakehemmerle** [comment]
Taking another look at this today with a bit more rust under my belt

TLDR dalek rejects verification when using the zero public key, while zebra allows this.

Next steps: 

1. Is it worth the overhead to check if its a zero key right before this: https://github.com/paritytech/substrate/blob/0ba251c9388452c879bfcca425ada66f1f9bc802/primitives/core/src/ed25519.rs#L515
and rejecting the signature.

2. or should I modify the assertion to accept a the zero key, and do a burn in?

If 2, do i just fork polkadot and replace everything that points to the paritytech/substrate repo to my repo/branch?

**Comment by jakehemmerle** [comment]
CC'ing @burdges @gilescope @bkchr on for the above, in case you don't get notifications on closed issues unless you're mentioned ^

**Comment by bkchr** [comment]
You can use [diener](https://github.com/bkchr/diener) to replace the branches etc to point to your own forks.

Regarding the zero public key, what does ed25519 standard says to this?

**Comment by jakehemmerle** [comment]
Cool tool!

I spent about 15 minutes looking for any sort of mention of how to handle zero keys (I found, literally, nothing), so I'm just going to do a burn in

**Comment by jakehemmerle** [comment]
Burn in passed. That made my day

I'm going to do #2: modify the assertion to accept the all-zero key

**Comment by jakehemmerle** [comment]
[Done](https://github.com/jakehemmerle/substrate/commit/34e895b2ffe09bf9669960932fdad140953244ff)

**Comment by burdges** [comment]
We do not care about the zero key verifying everything behavior.  Zero session or controller key are byzantine, while other zero account keys have no impact.

**Comment by jakehemmerle** [comment]
Cool, I'll put up a new PR shortly.

**Comment by bkchr** [comment]
> Ideally this PR should contain a label to indicate a change in the host function (for people still pending to do the dependency upgrade) (cc @bkchr )

We have done some sync testing on Polkadot/Kusama to ensure that nothing breaks. However, we should extend this process to Parachains as well.

**Comment by crystalin** [comment]
I think, no matter if it impacts an existing chain or not, when an assumption of a host function changes, it should be tagged, don't you think ?
